packages/native/src/NavigationContainer.tsx

```diff
@@ -46,13 +48,27 @@ const NavigationContainer = React.forwardRef(function NavigationContainer(
     ...linking,
   });
 
-  const [isReady, initialState] = useThenable(getInitialState);
+  const [isResolved, initialState] = useThenable(getInitialState);
 
   React.useImperativeHandle(ref, () => refContainer.current);
 
   const linkingContext = React.useMemo(() => ({ options: linking }), [linking]);
 
-  if (rest.initialState == null && isLinkingEnabled && !isReady) {
+  const isReady = rest.initialState != null || !isLinkingEnabled || isResolved;
+
+  const onReadyRef = React.useRef(onReady);
+
+  React.useEffect(() => {
+    onReadyRef.current = onReady;
+  });
+
+  React.useEffect(() => {
+    if (isReady) {
```

Comment:
shouldn't we prevent before calling it twice?

Reply:
I think it will be fine because `isReady` will only change once and we have it in dep array.